### PR TITLE
🌱 test/e2e/in-memory: improve performance by fixing locking issues

### PR DESCRIPTION
**Pull request by sbueringer**
Signed-off-by: Stefan Büringer buringerst@vmware.com

**What this PR does / why we need it**:
Currently the in-memory provider gets slow if we're getting to very high cluster numbers.

Basically the problem was that when we create a new apiserver we locked the global lock. This was slowing down all requests served by the apiserver because that lock is used during every tls handshake and in every request (to get the resourceGroup).

Kudos to @fabriziopandini for the collaboration on this.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Related to #8814

### CODE REVIEW

File: test/infrastructure/inmemory/internal/server/certs.go [hunk 1]
```diff
@@ -26,13 +26,19 @@ import (
 	"sigs.k8s.io/cluster-api/util/certs"
 )
 
-// newCertAndKey builds a new cert and key signed with the given CA and with the given config.
-func newCertAndKey(caCert *x509.Certificate, caKey *rsa.PrivateKey, config *certs.Config) (*x509.Certificate, *rsa.PrivateKey, error) {
-	key, err := certs.NewPrivateKey()
+var key *rsa.PrivateKey
+
+func init() {
```
Comment (sbueringer) [inline test/infrastructure/inmemory/internal/server/certs.go:31]: Generating a private key is slow. Now we only do it once on startup and share it across clusters

File: test/infrastructure/inmemory/internal/server/mux.go [hunk 2]
```diff
@@ -316,81 +318,104 @@ func (m *WorkloadClustersMux) initWorkloadClusterListenerWithPortLocked(wclName
...
-	var startErr error
-	startCh := make(chan struct{})
-	go func() {
-		startCh <- struct{}{}
-		if err := m.muxServer.ServeTLS(wcl.listener, "", ""); err != nil && !errors.Is(err, http.ErrServerClosed) {
-			startErr = err
-			m.log.Error(startErr, "Failed to start WorkloadClusterListener", "listenerName", wclName, "address", wcl.Address())
+	// Wait until the sever is working.
+	var pollErr error
+	err = wait.PollUntilContextTimeout(context.TODO(), 10*time.Millisecond, 1*time.Second, true, func(ctx context.Context) (done bool, err error) {
```
Comment (sbueringer) [inline test/infrastructure/inmemory/internal/server/mux.go:396]: With this change we wait until the listener serves (at least a tls handshake). This is quicker than always waiting 100 ms

**Review by fabriziopandini — COMMENTED** [review]
/lgtm
/approve

Thanks for moving forward this change!

**Pull request merged**